File: NodeJS/balanced-brackets.js
let s = "{{[[(())]]}}"

function isBalanced(s){
    let openingChars = ["{", "[", "("]
    let closingChars = ["}", "]", ")"]
    if(s.length % 2 != 0){
        return "NO"
    } else if(s.length == 0){
        return "YES"
    } else {
        for(let i = 0; i < s.length; i++){
            let j = closingChars.indexOf(s[i])
            if(j!=-1){
                let k = [...s.slice(0, i)].reverse().indexOf(openingChars[j])
                if(k==-1){
                    return "NO"
                } else {
                    if(isBalanced(s.slice(i-k, i))=="NO"){
                        return "NO"
                    } else {
                        return(isBalanced(s.slice(0, i-k-1) + s.slice(i+1)))
                    }
                }
            }
        }
        return "NO"
    }
}

console.log(isBalanced(s))

File: balanced-brackets.py
s = "{{[[(())]]}}"
'''
def isBalanced(s):
    print(f'Revisando {s}')
    openingChars = ["{", "[", "("]
    closingChars = ["}", "]", ")"]
    if len(s) % 2 != 0:
        return "NO"
    elif len(s)==0:
        return "YES"
    else:
        for i in range(len(s)):
            if s[i] in closingChars:
                j = closingChars.index(s[i])
                k = s[:i][::-1].find(openingChars[j])
                if k==-1:
                    return "NO"
                else:
                    if isBalanced(s[i-k:i])=="NO":
                        return "NO"
    return "YES"

print(isBalanced(s))
'''

def isBalanced(s):
    # print(f'Checking {s}')
    openingChars = ["{", "[", "("]
    closingChars = ["}", "]", ")"]
    if len(s) % 2 != 0:
        return "NO"
    elif len(s)==0:
        return "YES"
    else:
        for i in range(len(s)):
            if s[i] in closingChars:
                j = closingChars.index(s[i])
                k = s[:i][::-1].find(openingChars[j])
                if k==-1:
                    return "NO"
                else:
                    if isBalanced(s[i-k:i])=="NO":
                        return "NO"
                    else:
                        return isBalanced(s[:i-k-1] + s[i+1:])
        return "NO"

print(isBalanced(s))